scripts/stageapi/room/roomHandler.lua

```diff
@@ -1006,6 +1006,67 @@ function StageAPI.LoadGridsFromDataList(grids, gridInformation, entities, railsO
     return grids_spawned, minecart_points
 end
 
+StageAPI.AddCallback("StageAPI", "PRE_SPAWN_GRID", 0, function(gridEntry, gridInformation, entitySets, rng) --Handle rock replacements
+    if gridEntry.Type == GridEntityType.GRID_ROCK then
+        local currentRoom = StageAPI.GetCurrentRoom()
+        if not (gridEntry.SubType == 1 or currentRoom.Metadata:Has{Index = gridEntry.Index, Name = "PreventRandomization"}) then
+            local roll = StageAPI.Random(1, 100, rng)
+            local gridType
+            if shared.Room:GetTintedRockIdx() == gridEntry.Index then
+                if roll <= 5 then
+                    gridType = GridEntityType.GRID_ROCK_SS --Super Secret Rocks
+                else
+                    gridType = GridEntityType.GRID_ROCKT --Tinted Rocks
+                end
+            else
+                if roll <= 7 and roll > 3 then --Alt Rocks (Pots, Mushrooms, Skulls, etc.)
+                    gridType = GridEntityType.GRID_ROCK_ALT 
+            
+                elseif roll <= 3 and roll > 1 then  --Bomb Rocks 
+                    local gridPos = shared.Room:GetGridPosition(gridEntry.Index)
+                    local doorNearby
+                    for i = 0, DoorSlot.NUM_DOOR_SLOTS - 1 do --Don't spawn Bomb Rocks near doors
+                        if shared.Room:GetDoor(i) then
+                            local doorPos = room:GetDoorSlotPosition(i)
+                            local clampedPos = room:GetClampedPosition(doorPos, 20)
+                            if clampedPos.X == doorPos.X then 
+                                if math.abs(doorPos.X - gridPos.X) <= 40 and math.abs(doorPos.Y - gridPos.Y) <= 80 then
+                                    doorNearby = true
+                                    break
+                                end
+                            else 
+                                if math.abs(doorPos.Y - gridPos.Y) <= 40 and math.abs(doorPos.X - gridPos.X) <= 80 then
+                                    doorNearby = true
+                                    break
+                                end
+                            end
+                        end
+                    end
+
+                    if not doorNearby then
+                        gridType = GridEntityType.GRID_ROCK_BOMB
+                    end
+            
+                elseif roll < 1 then --Fool's Gold Rocks 
```

Comment:
Do fool's gold rocks ever spawn? I was pretty sure StageAPI.Random was an inclusive range, 1-100 including 1 and 100, and not 0.

It's probably not worth the trouble with the base API but I wonder if we should do anything about unlocks...

Reply:
Meant to make it <= 1 not < 1, fixed that

As for unlocks yeah Id rather not worry about that for now, going to wait for the API to be updated for that